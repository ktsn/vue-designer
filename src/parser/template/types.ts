import { Range } from '../modifier'

export type ElementChild = Element | TextNode | ExpressionNode

interface BaseNode extends Range {
  type: string
}

export interface Template extends BaseNode {
  type: 'Template'
  attrs: Record<string, Attribute>
  children: ElementChild[]
}

export interface Element extends BaseNode {
  type: 'Element'
  path: number[]
  name: string
  startTag: StartTag
  endTag?: EndTag
  children: ElementChild[]
}

export interface StartTag extends BaseNode {
  type: 'StartTag'

  // It simply maps attributes with the following rule:
  // 1. Static attributes are mapped to `attrs`
  // 2. Normal v-bind attributes are mapped to `props`
  // 3. v-bind attributes having `prop` modifier are mapped to `domProps`
  // 4. Other directives are in `directives`
  // `attrs` and `props` may include `class` and `style` unlike Vue.js template compiler.
  // Object style v-bind (v-bind="obj") is not supported yet. It will be in `directives`.
  attrs: Record<string, Attribute>
  props: Record<string, Directive>
  domProps: Record<string, Directive>
  directives: Directive[]

  selfClosing: boolean
}

export interface EndTag extends BaseNode {
  type: 'EndTag'
}

export interface TextNode extends BaseNode {
  type: 'TextNode'
  path: number[]
  text: string
}

export interface ExpressionNode extends BaseNode {
  type: 'ExpressionNode'
  path: number[]
  expression: string
}

export interface Attribute extends BaseNode {
  type: 'Attribute'
  attrIndex: number
  name: string
  value?: string
}

export interface Directive extends BaseNode {
  type: 'Directive'
  attrIndex: number
  name: string
  argument?: string
  modifiers: Record<string, boolean>
  expression?: string
  value?: any
}

export interface VForDirective extends Directive {
  name: 'for'
  left: string[]
  right?: string
}
