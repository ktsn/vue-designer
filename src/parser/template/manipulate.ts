import assert from 'assert'
import * as t from './types'
import { scopePrefix } from '../style/manipulate'
import { AssetResolver } from '../../asset-resolver'
import { clone } from '../../utils'

export function getNode(
  root: t.Template,
  path: number[]
): t.ElementChild | undefined {
  function loop(
    current: t.ElementChild,
    rest: number[]
  ): t.ElementChild | undefined {
    // If `rest` does not have any items,
    // `current` is the node we are looking for.
    if (rest.length === 0) {
      return current
    }

    // The current node does not have children,
    // then we cannot traverse any more.
    if (current.type !== 'Element') {
      return undefined
    }

    const next = current.children[rest[0]]
    if (!next) {
      return undefined
    } else {
      return loop(next, rest.slice(1))
    }
  }
  const [index, ...rest] = path
  const el = root.children[index]
  return el && loop(el, rest)
}

export function insertNode(
  root: t.Template,
  path: number[],
  el: t.ElementChild
): t.Template {
  function loop<T extends t.Element | t.Template>(
    parent: T,
    index: number,
    rest: number[]
  ): T {
    assert(index != null, '[template] index should not be null or undefined')

    const cs = parent.children

    // If `rest` is empty, insert the node to `index`
    if (rest.length === 0) {
      assert(
        0 <= index && index <= cs.length,
        "[template] cannot insert the node to '" +
          path.join('->') +
          "' as the last index is out of possible range: " +
          `0 <= ${index} <= ${cs.length}`
      )

      return clone(parent, {
        children: [...cs.slice(0, index), el, ...cs.slice(index)]
      })
    }

    const child = parent.children[index] as t.Element
    assert(
      child,
      "[template] cannot reach to the path '" +
        path.join('->') +
        "' as there is no node on the way"
    )
    assert(
      child.type === 'Element',
      "[template] cannot reach to the path '" +
        path.join('->') +
        "' as there is text or expression node on the way"
    )

    const [head, ...tail] = rest
    return clone(parent, {
      children: [
        ...cs.slice(0, index),
        loop(child, head, tail),
        ...cs.slice(index + 1)
      ]
    })
  }
  return loop(root, path[0], path.slice(1))
}

export function visitElements(
  node: t.Template,
  fn: (el: t.Element) => t.Element | void
): t.Template {
  function loop(node: t.ElementChild): t.ElementChild {
    switch (node.type) {
      case 'Element':
        const newNode = clone(node, {
          children: node.children.map(loop)
        })
        return fn(newNode) || newNode
      default:
        // Do nothing
        return node
    }
  }
  return clone(node, {
    children: node.children.map(loop)
  })
}

export function resolveAsset(
  template: t.Template,
  baseUrl: string,
  resolver: AssetResolver
): t.Template {
  return visitElements(template, el => {
    const src = el.startTag.attrs.src
    if (el.name === 'img' && src && src.value) {
      const resolvedSrc = clone(src, {
        value: resolver.pathToUrl(src.value, baseUrl)
      })

      return clone(el, {
        startTag: clone(el.startTag, {
          attrs: clone(el.startTag.attrs, {
            src: resolvedSrc
          })
        })
      })
    }
  })
}

export function addScope(node: t.Template, scope: string): t.Template {
  const scopeName = scopePrefix + scope

  return visitElements(node, el => {
    return clone(el, {
      startTag: clone(el.startTag, {
        attrs: clone(el.startTag.attrs, {
          [scopeName]: {
            type: 'Attribute',
            attrIndex: -1,
            name: scopeName,
            range: [-1, -1]
          }
        })
      })
    })
  })
}
