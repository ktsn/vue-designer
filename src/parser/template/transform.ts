import { AST } from 'vue-eslint-parser'
import * as t from './types'

type RootElement = AST.VElement & AST.HasConcreteInfo
type ChildNode = AST.VElement | AST.VText | AST.VExpressionContainer

export function transformTemplate(body: RootElement, code: string): t.Template {
  return {
    type: 'Template',
    range: body.range,
    attributes: transformAttributes(body.startTag.attributes),
    children: body.children.map((child, i) => transformChild(child, code, [i]))
  }
}

function transformElement(
  el: AST.VElement,
  code: string,
  path: number[]
): t.Element {
  const attrs = el.startTag.attributes

  const start = startTag(
    transformAttributes(attrs),
    transformDirectives(attrs, code),
    el.startTag.selfClosing,
    el.startTag.range
  )

  const end = el.endTag && endtag(el.endTag.range)

  return element(
    path,
    el.name,
    start,
    end || undefined,
    el.children.map((child, i) => transformChild(child, code, path.concat(i))),
    el.range
  )
}

function transformAttributes(
  attrs: (AST.VAttribute | AST.VDirective)[]
): Record<string, t.Attribute> {
  const res: Record<string, t.Attribute> = {}

  attrs.forEach((attr, index) => {
    if (attr.directive) {
      return
    }
    res[attr.key.name] = attribute(
      index,
      attr.key.name,
      attr.value ? attr.value.value : undefined,
      attr.range
    )
  })

  return res
}

function transformDirectives(
  attrs: (AST.VAttribute | AST.VDirective)[],
  code: string
): t.Directive[] {
  return attrs
    .map((attr, index) => {
      if (!attr.directive) {
        return null
      }

      if (attr.key.name === 'for') {
        return transformVForDirective(attr, index, code)
      } else {
        return transformDirective(attr, index, code)
      }
    })
    .filter((dir): dir is t.Directive => dir !== null)
}

function transformDirective(
  node: AST.VDirective,
  index: number,
  code: string
): t.Directive {
  const exp = node.value && node.value.expression
  const expStr = exp ? extractExpression(exp, code) : undefined
  return directive(
    index,
    node.key.name,
    node.key.argument || undefined,
    node.key.modifiers,
    expStr,
    node.range
  )
}

function transformVForDirective(
  node: AST.VDirective,
  index: number,
  code: string
): t.VForDirective {
  const exp =
    node.value &&
    node.value.expression &&
    node.value.expression.type === 'VForExpression'
      ? node.value.expression
      : null

  return vForDirective(
    index,
    exp ? exp.left.map(l => extractExpression(l, code)) : [],
    exp ? extractExpression(exp.right, code) : undefined,
    node.range
  )
}

function transformChild(
  child: ChildNode,
  code: string,
  path: number[]
): t.ElementChild {
  switch (child.type) {
    case 'VElement':
      return transformElement(child, code, path)
    case 'VText':
      return textNode(path, child.value, child.range)
    case 'VExpressionContainer':
      const exp = child.expression
      return expressionNode(
        path,
        exp ? extractExpression(exp, code) : '',
        child.range
      )
  }
}

function extractExpression(node: AST.HasLocation, code: string): string {
  return code.slice(node.range[0], node.range[1])
}

function element(
  path: number[],
  name: string,
  startTag: t.StartTag,
  endTag: t.EndTag | undefined,
  children: t.ElementChild[],
  range: [number, number]
): t.Element {
  return {
    type: 'Element',
    path,
    name,
    startTag,
    endTag,
    children,
    range
  }
}

function startTag(
  attributes: Record<string, t.Attribute>,
  directives: t.Directive[],
  selfClosing: boolean,
  range: [number, number]
): t.StartTag {
  return {
    type: 'StartTag',
    attributes,
    directives,
    selfClosing,
    range
  }
}

function endtag(range: [number, number]): t.EndTag {
  return {
    type: 'EndTag',
    range
  }
}

function textNode(
  path: number[],
  text: string,
  range: [number, number]
): t.TextNode {
  return {
    type: 'TextNode',
    path,
    text,
    range
  }
}

function expressionNode(
  path: number[],
  expression: string,
  range: [number, number]
): t.ExpressionNode {
  return {
    type: 'ExpressionNode',
    path,
    expression,
    range
  }
}

function attribute(
  attrIndex: number,
  name: string,
  value: string | undefined,
  range: [number, number]
): t.Attribute {
  return {
    type: 'Attribute',
    attrIndex,
    name,
    value,
    range
  }
}

function directive(
  attrIndex: number,
  name: string,
  argument: string | undefined,
  modifiers: string[],
  expression: string | undefined,
  range: [number, number]
): t.Directive {
  const mod: Record<string, boolean> = {}
  modifiers.forEach(m => {
    mod[m] = true
  })
  return {
    type: 'Directive',
    attrIndex,
    name,
    argument,
    modifiers: mod,
    expression,
    range
  }
}

function vForDirective(
  attrIndex: number,
  left: string[],
  right: string | undefined,
  range: [number, number]
): t.VForDirective {
  return {
    type: 'Directive',
    attrIndex,
    name: 'for',
    modifiers: {},
    left,
    right,
    range
  }
}
