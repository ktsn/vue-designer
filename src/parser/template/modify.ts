import { TETemplate, TETextNode, TEChild, TEElement } from './types'
import { getNode } from './manipulate'
import {
  insertBefore,
  singleIndentStr,
  insertAfter,
  Modifier,
  empty
} from '../modifier'

export function insertToTemplate(
  template: TETemplate,
  path: number[],
  value: string
): Modifier {
  const parentPath = path.slice(0, -1)
  const indent = inferTemplateIndentAt(template, parentPath)
  const target = getNode(template, path)
  if (target) {
    // If target path points an existing node, we can insert a new string before it.
    const post = '\n' + indent + singleIndentStr
    return insertBefore(target, value + post)
  }

  const index = path[path.length - 1]
  const last = getNode(template, parentPath.concat(index - 1))
  if (last) {
    // If target path points nothing, it indicates we will insert a new string
    // as the last element. If the current last node has indentation, we need
    // to add an extra indentation to align the children nodes and inserted one.
    //
    // ```
    // <div>
    //   <h1>Test</h1>
    // </div>
    // ```
    //
    // In the above case, the end tag has two spaces. If we want to insert a node
    // as the last child of `<div>` element, we will reuse the last text (indentation)
    // and add two spaces for indentation.
    //
    // ```
    // <div>
    //   <h1>Test</h1>
    //   <p>Hello</p>
    // </div>
    // ```
    //
    // But we should not add extra spaces if the end tag has no indentation. For example:
    //
    // ```
    // <div><strong>Test</strong></div>
    // ```
    //
    // In that case, we simply add a line break and appropriate indentation like:
    //
    // ```
    // <div><strong>Test</strong>
    //   <p>Hello</p>
    // </div>
    // ```
    const hasIndent = last.type === 'TextNode' && last.text.endsWith(indent)
    const pre = (hasIndent ? '' : '\n' + indent) + singleIndentStr
    const post = '\n' + indent
    return insertAfter(last, pre + value + post)
  }

  const parent = getNode(template, parentPath) as TEElement | undefined
  if (parent) {
    const pre = '\n' + indent + singleIndentStr
    const post = '\n' + indent
    return insertAfter(parent.startTag, pre + value + post)
  }

  return empty
}

function inferTemplateIndentAt(template: TETemplate, path: number[]): string {
  if (path.length === 0) {
    return ''
  }

  const parentPath = path.slice(0, -1)

  let node: TETextNode | undefined, iterating: TEChild | undefined
  for (let i = 0; (iterating = getNode(template, parentPath.concat(i))); i++) {
    if (iterating.type === 'TextNode') {
      node = iterating
      break
    }
  }

  if (!node) {
    return inferTemplateIndentAt(template, parentPath) + singleIndentStr
  }

  const match = /\n(\s+)/.exec(node.text)
  if (!match) {
    return ''
  }
  return match[1]
}
