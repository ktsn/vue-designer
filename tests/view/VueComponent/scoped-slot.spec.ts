import { createTemplate, h, render, a, exp } from '../../helpers/template'

describe('VueComponent scoped slot', () => {
  it('renders scoped slot content', () => {
    // prettier-ignore
    const template = createTemplate([
      h('Foo', [], [
        h('p', [a('slot-scope', 'props')], [exp('props.foo')])
      ])
    ])

    const components = {
      'file://Foo.vue': {
        // prettier-ignore
        template: createTemplate([
          h('div', [], [
            h('p', [], ['foo content']),
            h('slot', [a('foo', 'test')], [])
          ])
        ])
      }
    }

    const wrapper = render(
      template,
      [],
      [],
      [
        {
          name: 'Foo',
          uri: 'file://Foo.vue'
        }
      ],
      components
    )

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders named scoped slot content', () => {
    // prettier-ignore
    const template = createTemplate([
      h('Foo', [], [
        h('p', [a('slot-scope', 'props')], [exp('props.foo')]),
        h('p', [a('slot', 'test'), a('slot-scope', 'props')], [exp('props.foo')])
      ])
    ])

    const components = {
      'file://Foo.vue': {
        // prettier-ignore
        template: createTemplate([
          h('div', [], [
            h('p', [], ['foo content']),
            h('slot', [a('foo', 'test1')], []),
            h('slot', [a('name', 'test'), a('foo', 'test2')], [])
          ])
        ])
      }
    }

    const wrapper = render(
      template,
      [],
      [],
      [
        {
          name: 'Foo',
          uri: 'file://Foo.vue'
        }
      ],
      components
    )

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('resolves template element children as a slot', () => {
    // prettier-ignore
    const template = createTemplate([
      h('Foo', [], [
        h('template', [a('slot-scope', 'props')], [
          h('strong', [], ['named']),
          h('span', [], [exp('props.foo')])
        ])
      ])
    ])

    const components = {
      'file://Foo.vue': {
        // prettier-ignore
        template: createTemplate([
          h('div', [], [
            h('slot', [a('foo', 'test')], [])
          ])
        ])
      }
    }

    const wrapper = render(
      template,
      [],
      [],
      [
        {
          name: 'Foo',
          uri: 'file://Foo.vue'
        }
      ],
      components
    )

    expect(wrapper.html()).toMatchSnapshot()
  })
})
