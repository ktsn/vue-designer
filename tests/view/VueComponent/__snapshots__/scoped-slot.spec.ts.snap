// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VueComponent scoped slot > renders named scoped slot content 1`] = `
<div>
  <style></style>
  <div class="">
    <style></style>
    <div class="" data-scope-scope-id="">
      <p class="" data-scope-scope-id="">foo content</p>
      <p class="" slot-scope="props"><span class="">test1</span></p>
      <p class="" slot-scope="props"><span class="">test2</span></p>
    </div>
  </div>
</div>
`;

exports[`VueComponent scoped slot > renders scoped slot content 1`] = `
<div>
  <style></style>
  <div class="">
    <style></style>
    <div class="" data-scope-scope-id="">
      <p class="" data-scope-scope-id="">foo content</p>
      <p class="" slot-scope="props"><span class="">test</span></p>
    </div>
  </div>
</div>
`;

exports[`VueComponent scoped slot > resolves template element children as a slot 1`] = `
<div>
  <style></style>
  <div class="">
    <style></style>
    <div class="" data-scope-scope-id="">
      <div class="">named</div>
      <div class=""><span class="">test</span></div>
    </div>
  </div>
</div>
`;
